diff --git a/makechrootpkg b/makechrootpkg
index c9ae879..03ec607 100755
--- a/makechrootpkg
+++ b/makechrootpkg
@@ -28,13 +28,7 @@ export LANG=C
 
 shopt -s extglob
 
-# check if messages are to be printed using color
-if [[ -t 2 ]]; then
-	colorize
-else
-	# shellcheck disable=2034
-	declare -gr ALL_OFF='' BOLD='' BLUE='' GREEN='' RED='' YELLOW=''
-fi
+colorize
 
 stat_busy() {
 	local mesg=$1; shift
