diff --git a/makechrootpkg b/makechrootpkg
index c9ae879..03ec607 100755
--- a/makechrootpkg
+++ b/makechrootpkg
@@ -503,7 +497,7 @@ download_sources() {
 	# Ensure sources are downloaded
 	sudo -u "$makepkg_user" --preserve-env=GNUPGHOME \
 		env SRCDEST="$SRCDEST" BUILDDIR="$WORKDIR" \
-		makepkg --config="$copydir/etc/makepkg.conf" --verifysource -o "${verifysource_args[@]}" ||
+		"$copydir/bin/makepkg" --config="$copydir/etc/makepkg.conf" --verifysource -o "${verifysource_args[@]}" ||
 		die "Could not download sources."
 }
 
